DbClient - NullPointerException inTransaction with failing statement

[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"

## Environment Details
* Helidon Version: 2.0.1
* Helidon SE
* JDK version: 11.0.8+10
* OS: Archlinux

----------

## Problem Description
[//]: # "Describe the bug in detail highlighting current behavior vs expected behavior"

When using `inTransaction()`, if an error occurs the Single/Multi fails with a NullPointerException. 

[//]: # "State if the problem is easily reproducible or happens intermittently"

Fails every time.

[//]: # "Include stack traces or command outputs"

## Steps to reproduce
[//]: # "Step by step instructions to reproduce the problem"
```
        // Works fine
        client.execute(tx -> tx.createGet("select 1").execute()).get();
        // Throws java.util.concurrent.ExecutionException: org.postgresql.util.PSQLException
        client.execute(tx -> tx.createGet("select syntax error 1").execute()).get();
        // Throws java.util.concurrent.ExecutionException: java.lang.NullPointerException
        client.inTransaction(tx -> tx.createGet("select syntax error 1").execute()).get();
```
[//]: # "Provide sample code/application if relevant"
Complete stacktrace: 

```
java.util.concurrent.ExecutionException: java.lang.NullPointerException: The CompletionStage completed with a null value

	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2063)
	at io.helidon.common.reactive.Single.get(Single.java:608)
[redacted]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:686)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$201/0000000000000000.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall$$Lambda$202/0000000000000000.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$$Lambda$326/0000000000000000.apply(Unknown Source)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:212)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$352/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:208)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:137)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$274/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$273/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$272/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$278/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$274/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$273/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$272/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$278/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$274/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$273/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$272/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher$$Lambda$232/0000000000000000.accept(Unknown Source)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: java.lang.NullPointerException: The CompletionStage completed with a null value
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:73)
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:52)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:91)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:84)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152)
	at io.helidon.common.reactive.SingleToFuture.onError(SingleToFuture.java:71)
	at io.helidon.common.reactive.DeferredScalarSubscription.error(DeferredScalarSubscription.java:163)
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:67)
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:52)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:91)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:84)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152)
	at io.helidon.common.reactive.SingleToOptionalFuture.onError(SingleToOptionalFuture.java:70)
	at io.helidon.common.reactive.SingleFromPublisher$SingleSubscriber.onError(SingleFromPublisher.java:75)
	at io.helidon.common.reactive.SingleFlatMapMulti$FlatMapSubscriber$InnerSubscriber.onError(SingleFlatMapMulti.java:165)
	at io.helidon.common.reactive.SingleFlatMapMulti$FlatMapSubscriber$InnerSubscriber.onError(SingleFlatMapMulti.java:165)
	at io.helidon.common.reactive.SingleFlatMapMulti$FlatMapSubscriber.onError(SingleFlatMapMulti.java:102)
	at io.helidon.common.reactive.DeferredScalarSubscription.error(DeferredScalarSubscription.java:163)
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:67)
	at io.helidon.common.reactive.MultiFromCompletionStage$CompletionStageSubscription.accept(MultiFromCompletionStage.java:52)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:91)
	at io.helidon.common.reactive.MultiFromCompletionStage$AtomicBiConsumer.accept(MultiFromCompletionStage.java:84)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152)
	at io.helidon.dbclient.jdbc.JdbcStatementQuery.lambda$doExecute$2(JdbcStatementQuery.java:116)
	at io.helidon.dbclient.jdbc.JdbcStatementQuery$$Lambda$327/0000000000000000.run(Unknown Source)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:851)
```

--- a/dbclient/jdbc/pom.xml
+++ b/dbclient/jdbc/pom.xml
@@ -60,5 +60,10 @@
             <artifactId>hamcrest-all</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 </project>

--- a/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
+++ b/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
@@ -113,8 +113,7 @@ class JdbcDbClient implements DbClient {
         } else if (result instanceof Single) {
             Single<U> single = (Single<U>) result;
             return (T) single
-                    .flatMapSingle(it -> Single.create(execute.doCommit(), true)
-                            .map(conn -> it))
+                    .flatMapSingle(it -> Single.create(execute.doCommit().thenApply(unused -> it)))
                     .onError(RollbackHandler.create(execute, Level.WARNING)::apply);
         } else {
             execute.doRollback();

--- a/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
+++ b/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
@@ -113,13 +113,13 @@ class JdbcDbClient implements DbClient {
         } else if (result instanceof Single) {
             Single<U> single = (Single<U>) result;
             return (T) single
-                    .flatMapSingle(it -> Single.create(execute.doCommit())
+                    .flatMapSingle(it -> Single.create(execute.doCommit(), true)
                             .map(conn -> it))
                     .onError(RollbackHandler.create(execute, Level.WARNING)::apply);
         } else {
             execute.doRollback();
             throw new IllegalStateException("You must return a Single or Multi instance to inTransaction, yet "
-                    + "you provided: " + result.getClass().getName());
+                                                    + "you provided: " + result.getClass().getName());
         }
     }
 

--- a/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
+++ b/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
@@ -26,7 +26,6 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import io.helidon.common.mapper.MapperManager;
-import io.helidon.common.reactive.CompletionAwaitable;
 import io.helidon.common.reactive.Multi;
 import io.helidon.common.reactive.Single;
 import io.helidon.common.reactive.Subscribable;
@@ -113,17 +112,14 @@ class JdbcDbClient implements DbClient {
             return (T) multi;
         } else if (result instanceof Single) {
             Single<U> single = (Single<U>) result;
-
-            CompletionAwaitable<U> future = single.thenCompose(it -> execute.doCommit()
-                    .thenApply(conn -> it));
-
-            future.exceptionally(RollbackHandler.create(execute, Level.WARNING));
-
-            return (T) Single.create(future);
+            return (T) single
+                    .flatMapSingle(it -> Single.create(execute.doCommit())
+                            .map(conn -> it))
+                    .onError(RollbackHandler.create(execute, Level.WARNING)::apply);
         } else {
             execute.doRollback();
             throw new IllegalStateException("You must return a Single or Multi instance to inTransaction, yet "
-                                                    + "you provided: " + result.getClass().getName());
+                    + "you provided: " + result.getClass().getName());
         }
     }
 

--- a/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
+++ b/dbclient/jdbc/src/main/java/io/helidon/dbclient/jdbc/JdbcDbClient.java
@@ -117,7 +117,7 @@ class JdbcDbClient implements DbClient {
             CompletionAwaitable<U> future = single.thenCompose(it -> execute.doCommit()
                     .thenApply(conn -> it));
 
-            future = future.exceptionally(RollbackHandler.create(execute, Level.WARNING));
+            future.exceptionally(RollbackHandler.create(execute, Level.WARNING));
 
             return (T) Single.create(future);
         } else {

--- a/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
+++ b/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
@@ -18,6 +18,8 @@
 package io.helidon.dbclient.jdbc;
 
 import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
 
@@ -28,20 +30,29 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.is;
 
 import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 import org.mockito.Mockito;
 
 public class JdbcClient {
 
+    private static final ConnectionPool POOL = Mockito.mock(ConnectionPool.class);
+    private static final Connection CONN = Mockito.mock(Connection.class);
+    private static final PreparedStatement PREP_STATEMENT = Mockito.mock(PreparedStatement.class);
+
+    @BeforeAll
+    static void beforeAll() throws SQLException {
+        Mockito.when(CONN.prepareStatement("SELECT NULL FROM DUAL")).thenReturn(PREP_STATEMENT);
+        Mockito.when(POOL.connection()).thenReturn(CONN);
+        Mockito.when(PREP_STATEMENT.executeLargeUpdate()).thenReturn(0L);
+    }
+
     @Test
     void txErrorHandling() {
         String message = "BOOM IN TX!!!";
 
-        ConnectionPool cPool = Mockito.mock(ConnectionPool.class);
-        Mockito.when(cPool.connection()).thenReturn(Mockito.mock(Connection.class));
-
         JdbcDbClient dbClient = (JdbcDbClient) JdbcDbClientProviderBuilder.create()
-                .connectionPool(cPool)
+                .connectionPool(POOL)
                 .build();
         Object result = dbClient.inTransaction(tx -> Single.error(new RuntimeException(message)))
                 .onErrorResume(Function.identity())
@@ -49,5 +60,16 @@ public class JdbcClient {
 
         assertThat(result, CoreMatchers.instanceOf(RuntimeException.class));
         assertThat("Wrong exception propagated.", ((RuntimeException) result).getMessage(), is(equalTo(message)));
+    }
+
+    @Test
+    void txNullHandling() {
+        JdbcDbClient dbClient = (JdbcDbClient) JdbcDbClientProviderBuilder.create()
+                .connectionPool(POOL)
+                .build();
+        Object result = dbClient.inTransaction(tx -> tx.dml("SELECT NULL FROM DUAL"))
+                .await(200, TimeUnit.MILLISECONDS);
+
+        assertThat(result, is(equalTo(null)));
     }
 }

--- a/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
+++ b/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
@@ -48,6 +48,6 @@ public class JdbcClient {
                 .await(200, TimeUnit.MILLISECONDS);
 
         assertThat(result, CoreMatchers.instanceOf(RuntimeException.class));
-        assertThat("Wrong exception propagated.", ((RuntimeException) result).getCause().getMessage(), is(equalTo(message)));
+        assertThat("Wrong exception propagated.", ((RuntimeException) result).getMessage(), is(equalTo(message)));
     }
 }

--- a/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
+++ b/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
@@ -1,0 +1,53 @@
+/*
+ * Copyright (c)  2020 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.dbclient.jdbc;
+
+import java.sql.Connection;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import io.helidon.common.reactive.Single;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class JdbcClient {
+
+    @Test
+    void txErrorHandling() {
+        String message = "BOOM IN TX!!!";
+
+        ConnectionPool cPool = Mockito.mock(ConnectionPool.class);
+        Mockito.when(cPool.connection()).thenReturn(Mockito.mock(Connection.class));
+
+        JdbcDbClient dbClient = (JdbcDbClient) JdbcDbClientProviderBuilder.create()
+                .connectionPool(cPool)
+                .build();
+        Object result = dbClient.inTransaction(tx -> Single.error(new RuntimeException(message)))
+                .onErrorResume(Function.identity())
+                .await(200, TimeUnit.MILLISECONDS);
+
+        assertThat(result, CoreMatchers.instanceOf(RuntimeException.class));
+        assertThat("Wrong exception propagated.", ((RuntimeException) result).getCause().getMessage(), is(equalTo(message)));
+    }
+}

--- a/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
+++ b/dbclient/jdbc/src/test/java/io/helidon/dbclient/jdbc/JdbcClient.java
@@ -44,7 +44,7 @@ public class JdbcClient {
     static void beforeAll() throws SQLException {
         Mockito.when(CONN.prepareStatement("SELECT NULL FROM DUAL")).thenReturn(PREP_STATEMENT);
         Mockito.when(POOL.connection()).thenReturn(CONN);
-        Mockito.when(PREP_STATEMENT.executeLargeUpdate()).thenReturn(0L);
+        Mockito.when(PREP_STATEMENT.executeLargeUpdate()).thenReturn(1L);
     }
 
     @Test
@@ -63,13 +63,13 @@ public class JdbcClient {
     }
 
     @Test
-    void txNullHandling() {
+    void txResultHandling() {
         JdbcDbClient dbClient = (JdbcDbClient) JdbcDbClientProviderBuilder.create()
                 .connectionPool(POOL)
                 .build();
         Object result = dbClient.inTransaction(tx -> tx.dml("SELECT NULL FROM DUAL"))
                 .await(200, TimeUnit.MILLISECONDS);
 
-        assertThat(result, is(equalTo(null)));
+        assertThat(result, is(equalTo(1L)));
     }
 }

